tester.go

```diff
@@ -1,54 +1,94 @@
 package main
 
 import (
+	"context"
 	"fmt"
 	"net/http"
 	"time"
 )
 
-type Report struct {
-	Fail    uint64
-	Success uint64
-}
-
-func (r Report) String() string {
-	return fmt.Sprintf("Fail: %v, Success: %v", r.Fail, r.Success)
-}
+type requestChan chan struct{}
+type resultChan chan Report
 
 func WithRequests(u string, w uint64, n uint64) (Report, error) {
-	// TODO: parallelize with 'w' worker
-	var rep Report
 	if n < w {
-		return rep, fmt.Errorf("number of requests cannot be less then worker")
+		return Report{}, fmt.Errorf("number of requests cannot be less then worker")
 	}
-	for i := n; i > 0; i-- {
-		r, err := http.Get(u)
-		if err != nil {
-			rep.Fail++
-		} else {
-			rep.Success++
-			r.Body.Close()
-		}
+	req := make(requestChan, n)
+	res := make(resultChan)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for i := w; i > 0; i-- {
+		fmt.Println("M: start worker")
+		requestWorker(ctx, u, req, res)
 	}
+
+	var rep Report
+	for i := w; i > 0; i-- {
+		rep.Merge(<-res)
+	}
```

Comment:
This not only merges the report but perform a wait on the workers. Since we know how many requests are sent to the channel, you reuse this number (i) on the reception side. Can we come with a similar design that doesn't involve counting the number of elements which should have been sent (i)
Hint: closing a channel conveys the idea that nothing will ever been pushed again to the channel.

See: https://gobyexample.com/closing-channels and https://tour.golang.org/concurrency/4